src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -257,6 +258,338 @@ private struct IntPtrAndHandle
+
+        private static string MakeMarshalTypeRecord (Type type) {
+            return $"{{ \"marshalType\": {(int)GetMarshalTypeFromType(type)}, \"typePtr\": {type.TypeHandle.Value} }}";
+        }
+
+        private static MethodBase? MethodFromPointers (IntPtr typePtr, IntPtr methodPtr) {
+            if (methodPtr == IntPtr.Zero)
+                return null;
+
+            var tmp = new IntPtrAndHandle { ptr = methodPtr };
+            var methodHandle = tmp.handle;
+
+            if (typePtr != IntPtr.Zero) {
+                tmp.ptr = typePtr;
+                return MethodBase.GetMethodFromHandle(methodHandle, tmp.typeHandle);
+            } else {
+                return MethodBase.GetMethodFromHandle(methodHandle);
+            }
+        }
+
+        public static unsafe string? MakeMarshalSignatureInfo (IntPtr typePtr, IntPtr methodPtr) {
+            var mb = MethodFromPointers(typePtr, methodPtr);
+            if (mb == null)
+                return null;
+
+            var result = "{ " +
+                $"\"result\": {MakeMarshalTypeRecord((mb as MethodInfo)?.ReturnType ?? typeof(void))}, " +
+                $"\"typePtr\": {typePtr}, \"methodPtr\": {methodPtr}, " +
+                "\"parameters\": [";
+
+            int i = 0;
+            foreach (var p in mb.GetParameters()) {
+                if (i > 0)
+                    result += ", ";
+                result += MakeMarshalTypeRecord(p.ParameterType);
+                i++;
+            }
+
+            result += "] }";
+
+            return result;
+        }
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2070:UnrecognizedReflectionPattern",
+            Justification = "Trimming doesn't affect types eligible for marshalling. Different exception for invalid inputs doesn't matter.")]
+        private static unsafe string GetAndEscapeStringProperty (Type type, string name) {
+            var info = type.GetProperty(
+                name, BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic
+            );
+
+            var value = info?.GetValue(null) as string;
+            if (value == null)
+                return "null";
+
+            var result = "\"";
+            foreach (var ch in value) {
+                switch (ch) {
+                    case '\'':
+                        result += '\'';
+                        continue;
+                    case '"':
+                        result += '\"';
+                        continue;
+                    case '\\':
+                        result += "\\\\";
+                        continue;
+                    case '\n':
+                        result += "\\n";
+                        continue;
+                }
+
+                if (ch < ' ') {
+                    result += "\\x";
+                    if (ch <= 0xF)
+                        result += '0';
+                    result += ((int)ch).ToString("X");
+                } else {
+                    result += ch;
```

Comment:
As far as I know it's correct. The output is going to be UTF16 and it will be parsed as UTF16 by JSON.parse on the javascript side, which allows non-ascii characters in string literals. I would have preferred to use the BCL's JSON stuff but we can't realistically do that here

Reply:
GitHub won't let me reply to the below for some reason so I'll reply here

> We should also avoid building the string for the presumable default case where no escaping is needed.

I don't think we need to optimize this, because it's run once per type and then cached. I still will address all the very good code quality points here but I don't think we should waste time optimizing it or complicating the code with extra checks to do so.

If you want I can still optimize it to avoid building the string.